In [35]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

SyntaxError: invalid syntax (<ipython-input-35-897ff3ddd939>, line 19)

In [18]:
import json
w_data= 'C:/Users/mmishra30/Downloads/newyork_data.json'

## Load and explore the data

Next, let's load the data.

In [21]:
with open('C:/Users/mmishra30/Downloads/newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

Let's take a quick look at the data.

In [22]:
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

Notice how all the relevant data is in the features key, which is basically a list of the neighborhoods. So, let's define a new variable that includes this data.

In [24]:
neighborhoods_data = newyork_data['features']

Let's take a look at the first item in this list.

In [25]:
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

## Tranform the data into a pandas dataframe

The next task is essentially transforming this data of nested Python dictionaries into a pandas dataframe. So let's start by creating an empty dataframe.

In [26]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

Take a look at the empty dataframe to confirm that the columns are as intended.

In [27]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Then let's loop through the data and fill the dataframe one row at a time.

In [28]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

Quickly examine the resulting dataframe.

In [32]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

And make sure that the dataset has all 5 boroughs and 306 neighborhoods.

In [33]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


In [39]:
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim 

Solving environment: ...working... 
  - anaconda/win-64::ca-certificates-2020.10.14-0, anaconda/win-64::openssl-1.1.1h-he774522_0
  - anaconda/win-64::openssl-1.1.1h-he774522_0, defaults/win-64::ca-certificates-2020.10.14-0
  - anaconda/win-64::ca-certificates-2020.10.14-0, defaults/win-64::openssl-1.1.1h-he774522_0
  - defaults/win-64::ca-certificates-2020.10.14-0, defaults/win-64::openssl-1.1.1h-he774522_0done

# All requested packages already installed.



## Use geopy library to get the latitude and longitude values of New York City.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent ny_explorer, as shown below.

In [40]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


In [42]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.
Solving environment: ...working... 
  - anaconda/win-64::ca-certificates-2020.10.14-0
  - defaults/win-64::ca-certificates-2020.10.14-0done

## Package Plan ##

  environment location: C:\Users\mmishra30\Anaconda3

  added / updated specs:
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    altair-4.1.0               |             py_1         614 KB  conda-forge
    branca-0.4.2               |     pyhd8ed1ab_0          26 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    openssl-1.1.1h             |       he774522_0         5.8 MB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    ------------------------------------------------------------
                                         

## Create a map of New York with neighborhoods superimposed on top.

In [43]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

**Folium** is a great visualization library. Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Manhattan. So let's slice the original dataframe and create a new dataframe of the Manhattan data.

In [44]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

Let's get the geographical coordinates of Manhattan.

As we did with all of New York City, let's visualizat Manhattan the neighborhoods in it.

In [45]:
# create map of Manhattan using latitude and longitude values
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

## Define Foursquare Credentials and Version

In [65]:
CLIENT_ID = 'VEF5LC5RH2H1SCXAQ3VISVY1PHWMJWGFBERIHET4WWVCB5Z3' # your Foursquare ID
CLIENT_SECRET = 'DLA5KE21HCIHRI2MS3R3ILHSCDRBGY5JMOKRRPTV0KBOAYQ5' # your Foursquare Secret
ACCESS_TOKEN = 'IUOYRESKPVIPSWEQI3SWXYTU112Q40H1M45RZHJSIZ0IR0A5'  # your FourSquare Access Token
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: VEF5LC5RH2H1SCXAQ3VISVY1PHWMJWGFBERIHET4WWVCB5Z3
CLIENT_SECRET:DLA5KE21HCIHRI2MS3R3ILHSCDRBGY5JMOKRRPTV0KBOAYQ5


### Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [54]:
manhattan_data.loc[0, 'Neighborhood']

'Marble Hill'

In [174]:
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

In [173]:
manhattan_data.loc[1, 'Neighborhood']

'Chinatown'

In [48]:
neighborhood_latitude = manhattan_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = manhattan_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = manhattan_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Marble Hill are 40.87655077879964, -73.91065965862981.


### Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [175]:
## create URL
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# type your answer here


ACCESS_TOKEN = 'IUOYRESKPVIPSWEQI3SWXYTU112Q40H1M45RZHJSIZ0IR0A5'
 # create URL

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&oauth_token={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET,
    ACCESS_TOKEN, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL
## display URL


'https://api.foursquare.com/v2/venues/explore?&client_id=VEF5LC5RH2H1SCXAQ3VISVY1PHWMJWGFBERIHET4WWVCB5Z3&client_secret=DLA5KE21HCIHRI2MS3R3ILHSCDRBGY5JMOKRRPTV0KBOAYQ5&oauth_token=IUOYRESKPVIPSWEQI3SWXYTU112Q40H1M45RZHJSIZ0IR0A5&v=20180605&ll=40.87655077879964,-73.91065965862981&radius=500&limit=100'

Send the GET request and examine the resutls

In [176]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '603ecfd04a3f861bf26785a0'},
 'notifications': [{'type': 'notificationTray', 'item': {'unreadCount': 0}}],
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': '$-$$$$', 'key': 'price'},
    {'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Marble Hill',
  'headerFullLocation': 'Marble Hill, New York',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 68,
  'suggestedBounds': {'ne': {'lat': 40.88105078329964,
    'lng': -73.90471933917806},
   'sw': {'lat': 40.87205077429964, 'lng': -73.91659997808156}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b4429abf964a52037f225e3',
       'name': "Arturo's",
       'location': {'address': '5198 Broadway',
        'crossStreet': 'a

From the Foursquare lab in the previous module, we know that all the information is in the items key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [177]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a pandas dataframe.

In [178]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

nearby_venues.head()



<ipython-input-178-159db62baf83>:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  nearby_venues = json_normalize(venues) # flatten JSON


referralId  reasons.count  \
0  e-0-4b4429abf964a52037f225e3-0              0   
1  e-0-4baf59e8f964a520a6f93be3-1              0   
2  e-0-4b79cc46f964a520c5122fe3-2              0   
3  e-0-4b5357adf964a520319827e3-3              0   
4  e-0-55f81cd2498ee903149fcc64-4              0   

                                       reasons.items  \
0  [{'summary': 'This spot is popular', 'type': '...   
1  [{'summary': 'This spot is popular', 'type': '...   
2  [{'summary': 'This spot is popular', 'type': '...   
3  [{'summary': 'This spot is popular', 'type': '...   
4  [{'summary': 'This spot is popular', 'type': '...   

                   venue.id     venue.name venue.location.address  \
0  4b4429abf964a52037f225e3       Arturo's          5198 Broadway   
1  4baf59e8f964a520a6f93be3    Bikram Yoga          5500 Broadway   
2  4b79cc46f964a520c5122fe3  Tibbett Diner       3033 Tibbett Ave   
3  4b5357adf964a520319827e3        Dunkin'          5501 Broadway   
4  55f81cd2498ee903149fcc64      Starbucks         171 W 230th St   

  venue.location.crossStreet  venue.location.lat  venue.location.lng  \
0               at 225th St.           40.874412          -73.910271   
1               230th Street           40.876844          -73.906204   
2         btwn 230th & 231st           40.880404          -73.908937   
3                 W 230th St           40.877136          -73.906666   
4                Kimberly Pl           40.877531          -73.905582   

                       venue.location.labeledLatLngs  venue.location.distance  \
0  [{'label': 'display', 'lat': 40.87441177110231...                      240   
1  [{'label': 'display', 'lat': 40.87684369079793...                      376   
2  [{'label': 'display', 'lat': 40.8804044222466,...                      452   
3  [{'label': 'display', 'lat': 40.87713584201589...                      342   
4  [{'label': 'display', 'lat': 40.87753134921497...                      441   

  venue.location.postalCode venue.location.cc venue.location.city  \
0                     10463                US            New York   
1                     10463                US               Bronx   
2                     10463                US               Bronx   
3                     10463                US               Bronx   
4                     10463                US               Bronx   

  venue.location.state venue.location.country  \
0                   NY          United States   
1                   NY          United States   
2                   NY          United States   
3                   NY          United States   
4                   NY          United States   

                     venue.location.formattedAddress  \
0  [5198 Broadway (at 225th St.), New York, NY 10...   
1    [5500 Broadway (230th Street), Bronx, NY 10463]   
2  [3033 Tibbett Ave (btwn 230th & 231st), Bronx,...   
3      [5501 Broadway (W 230th St), Bronx, NY 10463]   
4    [171 W 230th St (Kimberly Pl), Bronx, NY 10463]   

                                    venue.categories venue.delivery.id  \
0  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...             72548   
1  [{'id': '4bf58dd8d48988d102941735', 'name': 'Y...               NaN   
2  [{'id': '4bf58dd8d48988d147941735', 'name': 'D...           2539986   
3  [{'id': '4bf58dd8d48988d148941735', 'name': 'D...           2200464   
4  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...               NaN   

                                  venue.delivery.url  \
0  https://www.seamless.com/menu/arturos-pizza-51...   
1                                                NaN   
2  https://www.seamless.com/menu/tibbett-diner-30...   
3  https://www.seamless.com/menu/dunkin-5501-broa...   
4                                                NaN   

  venue.delivery.provider.name       venue.delivery.provider.icon.prefix  \
0                     seamless  https://fastly.4sqi.net/img/general/cap/   
1                          NaN                           

In [152]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

nearby_venues.head()


venue.name                                   venue.categories  \
0       Arturo's  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...   
1    Bikram Yoga  [{'id': '4bf58dd8d48988d102941735', 'name': 'Y...   
2  Tibbett Diner  [{'id': '4bf58dd8d48988d147941735', 'name': 'D...   
3        Dunkin'  [{'id': '4bf58dd8d48988d148941735', 'name': 'D...   
4      Starbucks  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...   

   venue.location.lat  venue.location.lng  
0           40.874412          -73.910271  
1           40.876844          -73.906204  
2           40.880404          -73.908937  
3           40.877136          -73.906666  
4           40.877531          -73.905582

In [153]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

<ipython-input-153-561c05f0fdd1>:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  nearby_venues = json_normalize(venues) # flatten JSON


name   categories        lat        lng
0       Arturo's  Pizza Place  40.874412 -73.910271
1    Bikram Yoga  Yoga Studio  40.876844 -73.906204
2  Tibbett Diner        Diner  40.880404 -73.908937
3        Dunkin'   Donut Shop  40.877136 -73.906666
4      Starbucks  Coffee Shop  40.877531 -73.905582

In [154]:
nearby_venues

name  \
0                                            Arturo's   
1                                         Bikram Yoga   
2                                       Tibbett Diner   
3                                             Dunkin'   
4                                           Starbucks   
5                    Astral Fitness & Wellness Center   
6                                       Blink Fitness   
7                               Land & Sea Restaurant   
8                                           Starbucks   
9                           TCR The Club of Riverdale   
10                                           Rite Aid   
11                                          T.J. Maxx   
12                                     Vitamin Shoppe   
13                                             Subway   
14                                     Baskin-Robbins   
15                                      Boston Market   
16                                         Five Below   
17                                           GameStop   
18                                             Subway   
19                                    Parrilla Latina   
20                                      Auntie Anne's   
21                                 Lot Less Closeouts   
22                               The Children's Place   
23               America's Best Contacts & Eyeglasses   
24                                 CTown Supermarkets   
25                                            TD Bank   
26                                     Broadway Plaza   
27                                         Forever 21   
28                                     Planet Fitness   
29                                             Target   
30               U-Haul Moving & Storage of Riverdale   
31                                  Pizza Hut Express   
32                                        Lane Bryant   
33                                          Starbucks   
34            Metro North - Marble Hill Train Station   
35                                         Party City   
36                                          Marshalls   
37                                    Kam Wah Kitchen   
38                            Kennedy Chicken & Pizza   
39                                   Franco #2 Bakery   
40                                         Sally Food   
41                                Tavares Food Center   
42                                Taqueria Sinaloense   
43                                        Coffee Cart   
44                                          CA Liquor   
45                                     MTA - BX7/BX20   
46                             Marble Hill Playground   
47                                       Sprint Store   
48                Orange Park, Marble Hill, Bronx, NY   
49  MTA MaBSTOA Bus at W. 225th Street & Broadway ...   
50                             Applebee's Grill + Bar   
51                                         Deli-Dairy   
52                                         City Jeans   
53                                  Elyse Group 7 Inc   
54                                        Foot Locker   
55                   Country Delite: Milk Plant Store   
56                       Wolfeld Plastic Surgery, LLC   
57                           American Legion Post 774   
58                                        Mr Footwear   
59                                         Five Below   
60              Kingsbridge-Riverdale Farmers' Market   
61                                    California Farm   
62                                 Godwin Nails Salon   
63                                  Subway Sandwiches   
64                                    Grill 26 at TCR   
65                               Game & Comic Station   
66                   Kinhsbridge Quick Stop Deli Inc.   
67                                 Auntie Anne's #210   

                       categories        lat        lng  
0                     Pizza Place  40.874412 -73.910271  

And how many venues were returned by Foursquare?

In [179]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

68 venues were returned by Foursquare.


## 2. Explore Neighborhoods in Manhattan

Let's create a function to repeat the same process to all the neighborhoods in Manhattan


In [180]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now write the code to run the above function on each neighborhood and create a new dataframe called manhattan_venues.

In [181]:
#def getNearbyVenues(names, latitudes, longitudes, radius=500):
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude'])



Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


Let's check the size of the resulting dataframe

In [183]:
print(manhattan_venues.shape)
manhattan_venues.head()

(3182, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude          Venue  \
0  Marble Hill              40.876551               -73.91066       Arturo's   
1  Marble Hill              40.876551               -73.91066    Bikram Yoga   
2  Marble Hill              40.876551               -73.91066  Tibbett Diner   
3  Marble Hill              40.876551               -73.91066        Dunkin'   
4  Marble Hill              40.876551               -73.91066      Starbucks   

   Venue Latitude  Venue Longitude Venue Category  
0       40.874412       -73.910271    Pizza Place  
1       40.876844       -73.906204    Yoga Studio  
2       40.880404       -73.908937          Diner  
3       40.877136       -73.906666     Donut Shop  
4       40.877531       -73.905582    Coffee Shop

In [99]:
manhattan_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                
Battery Park City                       75                      75     75   
Carnegie Hill                           88                      88     88   
Central Harlem                          43                      43     43   
Chelsea                                100                     100    100   
Chinatown                              100                     100    100   
Civic Center                           100                     100    100   
Clinton                                100                     100    100   
East Harlem                             41                      41     41   
East Village                           100                     100    100   
Financial District                     100                     100    100   
Flatiron                               100                     100    100   
Gramercy                                83                      83     83   
Greenwich Village                      100                     100    100   
Hamilton Heights                        63                      63     63   
Hudson Yards                            70                      70     70   
Inwood                                  55                      55     55   
Lenox Hill                             100                     100    100   
Lincoln Square                          95                      95     95   
Little Italy                           100                     100    100   
Lower East Side                         47                      47     47   
Manhattan Valley                        38                      38     38   
Manhattanville                          45                      45     45   
Marble Hill                             24                      24     24   
Midtown                                100                     100    100   
Midtown South                          100                     100    100   
Morningside Heights                     42                      42     42   
Murray Hill                            100                     100    100   
Noho                                   100                     100    100   
Roosevelt Island                        27                      27     27   
Soho                                   100                     100    100   
Stuyvesant Town                         19                      19     19   
Sutton Place                           100                     100    100   
Tribeca                                 84                      84     84   
Tudor City                              77                      77     77   
Turtle Bay                             100                     100    100   
Upper East Side                        100                     100    100   
Upper West Side                         88                      88     88   
Washington Heights                      83                      83     83   
West Village                           100                     100    100   
Yorkville                              100                     100    100   

                     Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                          
Battery Park City                75               75              75  
Carnegie Hill                    88               88              88  
Central Harlem                   43               43              43  
Chelsea                         100              100             100  
Chinatown                       100              100             100  
Civic Center                    100              100             100  
Clinton                         100              100             100  
East Harlem                      41               41              41  
East Village                    100              100             100  
Finan

Let's find out how many unique categories can be curated from all the returned venues

In [184]:
print('There are {} uniques categories.'.format(len(manhattan_venues['Venue Category'].unique())))

There are 329 uniques categories.


In [185]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")
manhattan_onehot
# a# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood']

## 3. Analyze Each Neighborhood

## one hot encoding
One hot encoding is a process by which categorical variables are converted into a form that could be provided to ML algorithms to do a better job in prediction.

One hot encoding explained in an image

CompanyName	Categoricalvalue	Price		VW	Acura	Honda	Price
VW	    1	20000		                       1	0	0	   20000
Acura	2	10011		                       0	1	0	   10011
Honda	3	50000		                       0	0	1	   50000
Honda	3	10000		                       0	0	1	   10000


In [186]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Marble Hill                  0               0                  0   
1  Marble Hill                  0               0                  0   
2  Marble Hill                  0               0                  0   
3  Marble Hill                  0               0                  0   
4  Marble Hill                  0               0                  0   

   African Restaurant  American Restaurant  Antique Shop  Arepa Restaurant  \
0                   0                    0             0                 0   
1                   0                    0             0                 0   
2                   0                    0             0                 0   
3                   0                    0             0                 0   
4                   0                    0             0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auditorium  Australian Restaurant  \
0                 0                   0           0                      0   
1                 0                   0           0                      0   
2                 0                   0           0                      0   
3                 0                   0           0                      0   
4                 0                   0           0                      0   

   Austrian Restaurant  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                    0          0           0           0       0     0    0   
1                    0          0           0           0       0     0    0   
2                    0          0           0           0       0     0    0   
3                    0          0           0           0       0     0    0   
4                    0          0           0           0       0     0    0   

   Basketball Court  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0                 0         0            0           0              0   
1                 0         0            0           0              0   
2                 0         0            0           0              0   
3                 0         0            0           0              0   
4                 0         0            0           0              0   

   Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  Board Shop  \
0                         0          0           0       0           0   
1                         0          0           0       0           0   
2                         0          0           0       0           0   
3                         0          0           0       0           0   
4                         0          0           0       0           0   

   Boat or Ferry  Bookstore  Boutique  Boxing Gym  Brazilian Restaurant  \
0              0          0         0           0                     0   
1              0          0         0           0                     0   
2              0          0         0           0                     0   
3              0          0         0           0                     0   
4              0          0         0           0                     0   

   Breakfast Spot  Bridal Shop  Bridge  Bubble Tea Shop  Building  \
0               0            0       0                0         0   
1               0            0       0                0         0   
2               0            0       0                0         0   
3               0            0       0                0         0   
4               0  

And let's examine the new dataframe size.

In [187]:
manhattan_onehot.shape

(3182, 330)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [190]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean().reset_index()
manhattan_grouped

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0     Battery Park City           0.000000        0.000000               0.00   
1         Carnegie Hill           0.000000        0.000000               0.00   
2        Central Harlem           0.000000        0.000000               0.00   
3               Chelsea           0.000000        0.000000               0.00   
4             Chinatown           0.000000        0.000000               0.00   
5          Civic Center           0.000000        0.000000               0.00   
6               Clinton           0.000000        0.000000               0.00   
7           East Harlem           0.000000        0.000000               0.00   
8          East Village           0.000000        0.000000               0.00   
9    Financial District           0.000000        0.000000               0.00   
10             Flatiron           0.000000        0.000000               0.00   
11             Gramercy           0.000000        0.000000               0.00   
12    Greenwich Village           0.010000        0.000000               0.00   
13     Hamilton Heights           0.000000        0.016667               0.00   
14         Hudson Yards           0.000000        0.000000               0.00   
15               Inwood           0.000000        0.000000               0.00   
16           Lenox Hill           0.000000        0.000000               0.01   
17       Lincoln Square           0.000000        0.000000               0.00   
18         Little Italy           0.000000        0.000000               0.00   
19      Lower East Side           0.000000        0.000000               0.00   
20     Manhattan Valley           0.000000        0.000000               0.00   
21       Manhattanville           0.000000        0.000000               0.00   
22          Marble Hill           0.000000        0.000000               0.00   
23              Midtown           0.000000        0.000000               0.00   
24        Midtown South           0.000000        0.000000               0.00   
25  Morningside Heights           0.000000        0.000000               0.00   
26          Murray Hill           0.000000        0.000000               0.00   
27                 Noho           0.000000        0.000000               0.00   
28     Roosevelt Island           0.000000        0.000000               0.00   
29                 Soho           0.000000        0.000000               0.00   
30      Stuyvesant Town           0.000000        0.000000               0.00   
31         Sutton Place           0.000000        0.010000               0.00   
32              Tribeca           0.000000        0.000000               0.00   
33           Tudor City           0.000000        0.000000               0.00   
34           Turtle Bay           0.000000        0.000000               0.00   
35      Upper East Side           0.000000        0.000000               0.00   
36      Upper West Side           0.011364        0.000000               0.00   
37   Washington Heights           0.012195        0.000000               0.00   
38         West Village           0.010000        0.010000               0.00   
39            Yorkville           0.000000        0.000000               0.00   

    African Restaurant  American Restaurant  Antique Shop  Arepa Restaurant  \
0             0.000000             0.013158          0.00          0.000000   
1             0.000000             0.011628          0.00          0.000000   
2             0.069767             0.046512          0.00          0.000000   
3             0.000000             0.040000          0.00          0.000000   
4             0.000000             0.040000          0.00          0.000000   
5             0.000000             0.040000          0.01          0.000000   
6             0.000000             0.040000          0.00          0.000000   
7             0.000000             0.000000          0.00

Let's confirm the new size

In [189]:
manhattan_grouped.shape

(40, 330)

In [191]:
for hood in manhattan_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = manhattan_grouped[manhattan_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']

----Battery Park City----
----Carnegie Hill----
----Central Harlem----
----Chelsea----
----Chinatown----
----Civic Center----
----Clinton----
----East Harlem----
----East Village----
----Financial District----
----Flatiron----
----Gramercy----
----Greenwich Village----
----Hamilton Heights----
----Hudson Yards----
----Inwood----
----Lenox Hill----
----Lincoln Square----
----Little Italy----
----Lower East Side----
----Manhattan Valley----
----Manhattanville----
----Marble Hill----
----Midtown----
----Midtown South----
----Morningside Heights----
----Murray Hill----
----Noho----
----Roosevelt Island----
----Soho----
----Stuyvesant Town----
----Sutton Place----
----Tribeca----
----Tudor City----
----Turtle Bay----
----Upper East Side----
----Upper West Side----
----Washington Heights----
----West Village----
----Yorkville----


## Let's print each neighborhood along with the top 5 most common venues

In [192]:
num_top_venues = 5

for hood in manhattan_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = manhattan_grouped[manhattan_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Battery Park City----
            venue  freq
0     Coffee Shop  0.07
1            Park  0.05
2           Hotel  0.05
3  Clothing Store  0.05
4             Gym  0.05


----Carnegie Hill----
            venue  freq
0     Coffee Shop  0.08
1  Cosmetics Shop  0.05
2            Café  0.05
3     Yoga Studio  0.03
4       Bookstore  0.03


----Central Harlem----
                venue  freq
0  African Restaurant  0.07
1  Seafood Restaurant  0.05
2  Chinese Restaurant  0.05
3   French Restaurant  0.05
4      Cosmetics Shop  0.05


----Chelsea----
                 venue  freq
0          Art Gallery  0.06
1          Coffee Shop  0.05
2               Bakery  0.05
3  American Restaurant  0.04
4       Ice Cream Shop  0.03


----Chinatown----
                 venue  freq
0   Chinese Restaurant  0.08
1               Bakery  0.08
2  American Restaurant  0.04
3    Hotpot Restaurant  0.04
4         Cocktail Bar  0.04


----Civic Center----
                  venue  freq
0           Coffee Shop  0.08


### Let's put that into a pandas dataframe

First, let's write a function to sort the venues in descending order.

In [193]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [194]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = manhattan_grouped['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Battery Park City           Coffee Shop                 Hotel   
1      Carnegie Hill           Coffee Shop        Cosmetics Shop   
2     Central Harlem    African Restaurant        Cosmetics Shop   
3            Chelsea           Art Gallery                Bakery   
4          Chinatown                Bakery    Chinese Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Park                   Gym        Clothing Store   
1                  Café           Yoga Studio           Pizza Place   
2                   Bar     French Restaurant   American Restaurant   
3           Coffee Shop   American Restaurant                 Hotel   
4          Cocktail Bar   American Restaurant     Hotpot Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Boat or Ferry         Memorial Site          Burger Joint   
1             Wine Shop                   Gym     French Restaurant   
2    Chinese Restaurant    Seafood Restaurant             Bookstore   
3        Ice Cream Shop    Seafood Restaurant    Italian Restaurant   
4          Dessert Shop    Salon / Barbershop                   Spa   

  9th Most Common Venue 10th Most Common Venue  
0                 Plaza            Beer Garden  
1             Bookstore           Cocktail Bar  
2  Ethiopian Restaurant       Tapas Restaurant  
3     French Restaurant                 Market  
4          Optical Shop     Mexican Restaurant

## 4. Cluster Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [195]:
# set number of clusters
kclusters = 5

manhattan_grouped_clustering = manhattan_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(manhattan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 0, 0, 0, 1, 1, 3, 0, 1])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [196]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

manhattan_merged = manhattan_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

manhattan_merged.head() # check the last columns!

Borough        Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Manhattan         Marble Hill  40.876551 -73.910660               1   
1  Manhattan           Chinatown  40.715618 -73.994279               0   
2  Manhattan  Washington Heights  40.851903 -73.936900               0   
3  Manhattan              Inwood  40.867684 -73.921210               3   
4  Manhattan    Hamilton Heights  40.823604 -73.949688               3   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0        Discount Store           Coffee Shop        Sandwich Place   
1                Bakery    Chinese Restaurant          Cocktail Bar   
2                  Café                Bakery                  Bank   
3    Mexican Restaurant                Lounge            Restaurant   
4           Pizza Place           Coffee Shop                  Café   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                   Gym           Yoga Studio        Tennis Stadium   
1   American Restaurant     Hotpot Restaurant          Dessert Shop   
2     Mobile Phone Shop    Italian Restaurant         Deli / Bodega   
3                  Café    Chinese Restaurant              Wine Bar   
4    Mexican Restaurant         Deli / Bodega           Yoga Studio   

       7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0              Deli / Bodega      Department Store        Ice Cream Shop   
1         Salon / Barbershop                   Spa          Optical Shop   
2         Spanish Restaurant        Sandwich Place           Pizza Place   
3              Deli / Bodega           Pizza Place                Bakery   
4  Latin American Restaurant          Liquor Store                  Park   

  10th Most Common Venue  
0               Pharmacy  
1     Mexican Restaurant  
2                   Park  
3                   Park  
4           Cocktail Bar

Finally, let's visualize the resulting clusters

In [197]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(manhattan_merged['Latitude'], manhattan_merged['Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

### Cluster 1

In [198]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue    2nd Most Common Venue  \
1            Chinatown                Bakery       Chinese Restaurant   
2   Washington Heights                  Café                   Bakery   
5       Manhattanville    Seafood Restaurant              Coffee Shop   
6       Central Harlem    African Restaurant           Cosmetics Shop   
9            Yorkville    Italian Restaurant              Coffee Shop   
10          Lenox Hill    Italian Restaurant              Pizza Place   
12     Upper West Side                  Café       Italian Restaurant   
13      Lincoln Square                 Plaza                     Café   
17             Chelsea           Art Gallery                   Bakery   
18   Greenwich Village    Italian Restaurant           Clothing Store   
19        East Village                   Bar              Pizza Place   
20     Lower East Side    Chinese Restaurant                 Pharmacy   
22        Little Italy                Bakery                     Café   
23                Soho        Clothing Store       Italian Restaurant   
24        West Village    Italian Restaurant  New American Restaurant   
25    Manhattan Valley    Mexican Restaurant              Pizza Place   
31                Noho    Italian Restaurant              Art Gallery   

   3rd Most Common Venue     4th Most Common Venue  5th Most Common Venue  \
1           Cocktail Bar       American Restaurant      Hotpot Restaurant   
2                   Bank         Mobile Phone Shop     Italian Restaurant   
5       Sushi Restaurant        Mexican Restaurant     Chinese Restaurant   
6                    Bar         French Restaurant    American Restaurant   
9                    Bar                       Gym       Sushi Restaurant   
10           Coffee Shop              Cocktail Bar       Sushi Restaurant   
12                Bakery                       Bar               Wine Bar   
13          Concert Hall                   Theater  Performing Arts Venue   
17           Coffee Shop       American Restaurant                  Hotel   
18      Sushi Restaurant       American Restaurant      Indian Restaurant   
19          Cocktail Bar         Korean Restaurant     Italian Restaurant   
20                  Café               Pizza Place                 Bakery   
22       Bubble Tea Shop              Cocktail Bar     Chinese Restaurant   
23           Coffee Shop  Mediterranean Restaurant               Boutique   
24          Cocktail Bar       American Restaurant                   Park   
25                   Bar         Indian Restaurant            Coffee Shop   
31           Pizza Place         French Restaurant            Coffee Shop   

   6th Most Common Venue 7th Most Common Venue     8th Most Common Venue  \
1           Dessert Shop    Salon / Barbershop                       Spa   
2          Deli / Bodega    Spanish Restaurant            Sandwich Place   
5     Italian Restaurant      Ramen Restaurant          Cuban Restaurant   
6     Chinese Restaurant    Seafood Restaurant                 Bookstore   
9          Deli / Bodega             Wine Shop       Japanese Restaurant   
10          Burger Joint  Gym / Fitness Center                       Gym   
12           Pizza Place     Indian Restaurant               Coffee Shop   
13   Indie Movie Theater    Italian Restaurant                    Bakery   
17        Ice Cream Shop    Seafood Restaurant        Italian Restaurant   
18                   Gym              Boutique              Dessert Shop   
19              Wine Bar           Coffee Shop        Mexican Restaurant   
20           Art Gallery           Yoga Studio       Filipino Restaurant   
22           Coffee Shop              Tea Room  Mediterranean Restaurant   
23                Bakery    Seafood Restaurant        Salon / Barbershop   
24        Ice Cream Shop              Wine Bar                   Theater   
25         Grocery Store   Peruvian Restaurant              Noodle House   
31              

#### Cluster 2

In [199]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue    2nd Most Common Venue  \
0          Marble Hill        Discount Store              Coffee Shop   
8      Upper East Side               Exhibit       Italian Restaurant   
14             Clinton    Italian Restaurant     Gym / Fitness Center   
15             Midtown                 Hotel           Clothing Store   
16         Murray Hill           Coffee Shop                    Hotel   
21             Tribeca                  Park       Italian Restaurant   
27            Gramercy                   Bar       Italian Restaurant   
28   Battery Park City           Coffee Shop                    Hotel   
29  Financial District           Coffee Shop              Pizza Place   
30       Carnegie Hill           Coffee Shop           Cosmetics Shop   
32        Civic Center           Coffee Shop                      Spa   
34        Sutton Place    Italian Restaurant              Coffee Shop   
35          Turtle Bay    Italian Restaurant              Coffee Shop   
38            Flatiron    Italian Restaurant  New American Restaurant   
39        Hudson Yards  Gym / Fitness Center       Italian Restaurant   

   3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
0         Sandwich Place                     Gym           Yoga Studio   
8            Coffee Shop                  Bakery  Gym / Fitness Center   
14        Sandwich Place             Coffee Shop               Theater   
15               Theater     Sporting Goods Shop                   Spa   
16        Sandwich Place                     Bar   Japanese Restaurant   
21   American Restaurant             Coffee Shop              Wine Bar   
27           Pizza Place              Bagel Shop            Playground   
28                  Park                     Gym        Clothing Store   
29                   Bar            Cocktail Bar    Italian Restaurant   
30                  Café             Yoga Studio           Pizza Place   
32   American Restaurant            Cocktail Bar                 Hotel   
34                  Park    Gym / Fitness Center                   Gym   
35      Sushi Restaurant     Japanese Restaurant                  Park   
38   Japanese Restaurant  Furniture / Home Store                   Spa   
39   American Restaurant                    Café                 Hotel   

       6th Most Common Venue 7th Most Common Venue   8th Most Common Venue  \
0             Tennis Stadium         Deli / Bodega        Department Store   
8        American Restaurant                   Spa       French Restaurant   
14       American Restaurant                   Gym            Cocktail Bar   
15       American Restaurant           Coffee Shop                  Bakery   
16       American Restaurant          Burger Joint    Gym / Fitness Center   
21                       Spa                  Café          Scenic Lookout   
27        Mexican Restaurant           Coffee Shop  Thrift / Vintage Store   
28             Boat or Ferry         Memorial Site            Burger Joint   
29                Steakhouse                  Café     American Restaurant   
30                 Wine Shop                   Gym       French Restaurant   
32      Gym / Fitness Center     French Restaurant                    Park   
34    Furniture / Home Store   American Restaurant             Pizza Place   
35             Deli / Bodega    Seafood Restaurant        Ramen Restaurant   
38  Mediterranean Restaurant  Gym / Fitness Center             Coffee Shop   
39              Burger Joint            Restaurant             Coffee Shop   

   9th Most Common Venue 10th Most Common Venue  
0         Ice Cream Shop               Pharmacy  
8                  Hotel              Juice Bar  
14                   Spa              Wine Shop  
15        Sandwich Place              Bookstore  
16                Bakery      Jewish Restaurant  
21            Steakhouse             Poke Place  
27   American Restaurant         Pilates Studio  
28     

#### Cluster 3



In [200]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
37  Stuyvesant Town                  Park         Boat or Ferry   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
37           Coffee Shop                   Bar       Harbor / Marina   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37           Pet Service           Gas Station     German Restaurant   

   9th Most Common Venue 10th Most Common Venue  
37        Farmers Market   Gym / Fitness Center

#### Cluster 4



In [201]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 3, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
3                Inwood    Mexican Restaurant                Lounge   
4      Hamilton Heights           Pizza Place           Coffee Shop   
7           East Harlem    Mexican Restaurant                Bakery   
11     Roosevelt Island                  Park     Food & Drink Shop   
26  Morningside Heights                  Park             Bookstore   
36           Tudor City                  Park    Mexican Restaurant   

                       3rd Most Common Venue 4th Most Common Venue  \
3                                 Restaurant                  Café   
4                                       Café    Mexican Restaurant   
7                            Thai Restaurant         Deli / Bodega   
11  Residential Building (Apartment / Condo)        Sandwich Place   
26                       American Restaurant           Coffee Shop   
36                                      Café                 Diner   

        5th Most Common Venue 6th Most Common Venue  \
3          Chinese Restaurant              Wine Bar   
4               Deli / Bodega           Yoga Studio   
7   Latin American Restaurant        Sandwich Place   
11                Supermarket           Coffee Shop   
26                       Café          Burger Joint   
36              Deli / Bodega                   Gym   

        7th Most Common Venue 8th Most Common Venue  9th Most Common Venue  \
3               Deli / Bodega           Pizza Place                 Bakery   
4   Latin American Restaurant          Liquor Store                   Park   
7                        Café          Liquor Store             Taco Place   
11               Liquor Store          Soccer Field       Greek Restaurant   
26             Sandwich Place    Mexican Restaurant           Tennis Court   
36                Coffee Shop      Greek Restaurant  Vietnamese Restaurant   

   10th Most Common Venue  
3                    Park  
4            Cocktail Bar  
7             Gas Station  
11         Farmers Market  
26           Liquor Store  
36                 Garden

#### Cluster 5



In [202]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 4, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
33  Midtown South     Korean Restaurant                 Hotel   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
33             Hotel Bar   Japanese Restaurant          Dessert Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
33        Cosmetics Shop           Coffee Shop    Italian Restaurant   

   9th Most Common Venue 10th Most Common Venue  
33             Gastropub   Gym / Fitness Center